// vk-sound-player/dist/demo/audioModelDemo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../src/audioModel/index");
window.onload = function () {
    let playStatus = true;
    function a() {
        playStatus = !playStatus;
        let text = playStatus ? '播放' : '暂停1111';
        $('.play-btn').html(text);
    }
    // http://sc1.111ttt.cn/2015/1/07/15/100151835043
    let audio = index_1.renderAudio({
        url: '1675.mp3',
        endedFn: () => {
            $('.play-btn').html('播放');
        },
        audioStatusChange: a,
        autoPlay: true,
        progressData: {
            showProgress: true,
            parent: '.progress',
            style: {
                wrapper: {
                    width: "100%",
                    background: '#00d',
                    'border-radius': '0',
                }
            }
        },
        volumeData: {
            showVolume: true,
            parent: '.volumn-container',
            label: ".volume-label",
            style: {
                current: {
                    'width': '100%'
                }
            }
        },
        canplay: () => {
            alert('canplay');
        },
        loadedmetadat: () => {
            alert('loadedmetadat');
        },
        waiting: () => {
            alert('witting');
        },
        errorFn: () => {
            alert('error');
        }
    });
    console.log(audio);
    // oncanplay，onloadedmetadata，onwaiting，onerror
    $('.play-btn').on('click', function () {
        audio.togglePlay(playStatus);
    });
};
