// vk-runtime/src/pages/common/course/classroomChannel.js
import axios from 'axios'
import cookies from "cookies-js"
/**
 * params {
 *  studentId: 学生id
 *  source： 来源  家长端： PARENT   学习中心：LEARNING_CENTER
 * }
 *  考虑后期url 问题。 直接使用全路径
 */
function channel(classId, params) {
  let {studentId, source} = params
  let learningUrl = getUrl('learning.vipkid')
  let wopUrl = getUrl('www.vipkid')
  // let studentUrl = getUrl(`classroom-gateway.vipkid`) // classroom-gateway 目前使用全路径 如有pass配置， 修改下代码
  // 目前是配置 pass /rest/classroomgw
  let url = `/rest/classroomgw/api/class/${classId}/student/${studentId}/classroom/channel`
  return axios.get(url, {
    params: {
      source
    },
    timeout: 3000
  }).then(({ data }) => {
    if (data.code === 0 && data.msg === 'success') {
      return data.data.channel === 'WOP' ? `${location.protocol}//${wopUrl}/warrior/#/wop?classId=${classId}` : `${location.protocol}//${learningUrl}/classroom/major?onlineClassId=${classId}`
    } else {
      return `${location.protocol}//${learningUrl}/classroom/major?onlineClassId=${classId}`
    }
  }).catch(err => {
    console.error(err)
    return `${location.protocol}//${learningUrl}/classroom/major?onlineClassId=${classId}`
  })
}

const getUrl = (url) => {
  let host = location.host
  let reg = /^\w+\d+/
  let serverName = reg.exec(host) ? reg.exec(host)[0] : ''
  // 开发环境判断 开发环境使用自定义域名
  // if (host.indexOf('-qa') > -1) {
  //   serverName = 'a23'
  // }
  url = serverName ? `${serverName}-${url}-qa.com.cn` : host.indexOf('pre-') > -1 ? `pre-${url}.com.cn` : `${url}.com.cn`
  return url
}

function getChannel(onlineClassId) {
  let params = {
    studentId: cookies.get("studentId"),
    source: 'LEARNING_CENTER'
  }
  return channel(onlineClassId, params)
}
export default {
  getChannel
}
