// vk-runtime/webpack.dev.js
'use strict'
const webpack = require('webpack')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const path = require('path')
const ExtractTextPlugin = require('extract-text-webpack-plugin')
const CopyWebpackPlugin = require('copy-webpack-plugin')

// webpack bundle 分析
// const Visualizer = require('webpack-visualizer-plugin')
// const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin
module.exports = {
  output: {
    path: path.join(__dirname, "./output"),
    filename: '[name].js',
    chunkFilename: '[name].bundle.js',
    publicPath: '/'
  },
  module: {},
  plugins: [
    // 包分析插件
    // new BundleAnalyzerPlugin({
    //   reportFilename: './bundle.html'
    // }),
    // new Visualizer({
    //   filename: './statistics.html'
    // }),
    //复制营销活动插件
    // new CopyWebpackPlugin(
    //   [{
    //     from: 'star_lottery',
    //     to: './star_lottery/'
    //   }], {
    //     force: true,
    //     ignore: [],
    //     copyUnmodified: true,
    //   }),
    new HtmlWebpackPlugin({
      template: './src/index.html',
      filename: 'index.html',
      favicon: path.resolve('./src/img/favicon.png')
    }),
    //去除hash，用于开发时不会出现多个文件
    new ExtractTextPlugin({
      filename: 'src/css/style.css'
    }),
  ],
  devtool: 'source-map'
}
