// vk-runtime/src/pages/common/plugins/WorkEntrance.js
//处理课前课后资源跳转
import dragonUtils from '../../../utils/_dragon.js'
import axios from 'axios'
import cookies from 'cookies-js'

export default {
  install (Vue, options) {
    Vue.prototype.$getWorkUrl = (obj, context, type, clickId) => {
      axios.get('/rest/learninggw/api/pc/material/extra/getStudentExtraRoute', {
        params: {
          studentId: cookies.get("studentId"),
          onlineClassId: obj.onlineClassId,
          lessonId: obj.lessonId,
          courseId: obj.courseId,
          curriculumVersion: obj.curriculumVersion ? obj.curriculumVersion : 1,
          source: dragonUtils.isOnDragonClient() ? 2 : 1, //1PC 2PCAPP
          type: type == 'PRE' ? 1 : 2
        }
      }).then((res) => {
        let data = res.data.data
        if (data.isLc) {
          context.$router.push({path: data.url})
        } else {
          window.location.href = data.url + "&back_url=" + encodeURIComponent(document.location.href) //@back_url 用于作业端返回
        }
      })
      sa.track('learning_click', {
        'click_id': clickId + '_' + type.toLowerCase(),
        'online_class_id': obj.onlineClassId,
        'lesson_id': obj.lessonId,
        'class_id': obj.courseId //由于神策新增字段申请流程繁琐，先用class_id代表courseId
      })
    }
  }
}
