// egret-dino/webpack.dev.js
const merge = require('webpack-merge');
const common = require('./webpack.common.js');
const webpack = require('webpack');
const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const path = require("path");
module.exports = merge(common, {
  entry: {
    app: "./src/main"
  },
  output: {
    filename: '[name].bundle.js',
    path: path.resolve(__dirname, 'dist')
  },
  devtool: 'inline-source-map',
  plugins: [
    new webpack.IgnorePlugin(/vertx/),
    new webpack.NamedModulesPlugin(),
    new webpack.HotModuleReplacementPlugin(),
    new UglifyJSPlugin({
      uglifyOptions: {
        output: {
          comments: false,
          beautify: false
        },
        warnings: false,
        compress: true
      }
    }),
    new HtmlWebpackPlugin({
      title: 'egret for dino',
      template: './index.html'
    })
  ],
  devServer: {
    contentBase: './',
    host: '0.0.0.0',
    hot: true
  },
  //4.0配置
  optimization: {
    splitChunks: {
      cacheGroups: {
        commons: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendor",
          chunks: "all"
        }
      }
    }
  },
});
