// egret-dino/webpack.common.js
const path = require("path");
function resolve(dir) {
  return path.join(__dirname, '', dir)
}

module.exports = {
  resolve: {
    extensions: ['.js', '.json', '.ts'],
    alias: {
      '@': resolve('src'),
      '~': path.resolve(__dirname)
    }
  },

  module: {
    rules: [
      {
        test: /\.(gif|jpg|jpeg|png|svg)$/,//图片各类格式
        use: [
          {
            loader: 'url-loader',
            options: {
              limit: 1024,//图片大小
              name: '[name].[ext]'//图片名称规则
            }
          }
        ]
      },
      {
        test: /\.js$/,
        loader: 'babel-loader',
        exclude: /node_modules/
      },
      {
        test: /\.ts$/,
        loader: 'awesome-typescript-loader',
        query: {
          'doTypeCheck': true
        },
        include: path.resolve(__dirname, 'src'),
        exclude: /node_modules/
      },
      {
        type: 'javascript/auto',
        test: /\.(json)/,
        exclude: /(node_modules|bower_components)/,
        use: [{
          loader: 'file-loader',
          options: { name: '[name].[ext]' },
        }],
      },
      {
        test: /\.html$/,
        loader: 'html-loader'
      }
    ]
  }
};
