// vk-runtime/src/pages/project_demo/js/toolkit.js
export default class Toolkit {

  static fullTime(t){
    return t < 10 ? '0'+t : t
  }

  static getDaysAndHours(time){
    let day = fullTime(parseInt(time / (1000 * 60 * 60 * 24))).toString().split('')
    let hour = fullTime(parseInt((time % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60))).toString().split('')
    return {'day':day,'hour':hour}
  }

  static getHours(time){
    let timestamp = new Date().getTime()
    let _time = timestamp-time
    let day = parseInt(_time / (1000 * 60 * 60 * 24))
    let hour = parseInt((_time % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60))
    return day*24+hour
  }

  static getTime(time){
    let timestamp = new Date().getTime()
    let _time = timestamp-time
    let month = parseInt(_time / (1000 * 60 * 60 * 24 * 30))
    if (month > 0) {
      if (month == 1) {
        return month+' month ago'
      }else{
        return month+' months ago'
      } 
    }
    let day = parseInt(_time / (1000 * 60 * 60 * 24))
    if (day> 0) {
      if (day == 1) {
        return day+' day ago'
      }else{
        return day+' days ago'
      }
    }
    let hour = parseInt(_time / (1000 * 60 * 60))
    if (hour > 0) {
      if (hour == 1) {
        return hour+' hour ago'
      }else{
        return hour+' hours ago'
      } 
    }
    let min = parseInt(_time / (1000 * 60))
    if (min > 0) {
      return min+' min. ago'
    }

    let sec = parseInt(_time / (1000 * 60))
    if (sec > 0) {
      return sec+' sec. ago'
    }

    return ''
  }

  static formatData(timestamp){
    let time = new Date(timestamp)
    let y = time.getFullYear()
    let m = time.getMonth()+1
    let d = time.getDate()
    return y+'.'+fullTime(m)+'.'+fullTime(d)
  }

  static heartBeat() {
    dispatchForIpad('vkappbridge://ipadcommon/postHeartbeat')
  }
  static dispatchForIpad(src) {
    let _iframe = document.createElement('iframe')
    _iframe.setAttribute('src', src)
    _iframe.setAttribute('style', 'display:none;')
    document.body.appendChild(_iframe)
    _iframe.parentNode.removeChild(_iframe)
    _iframe = null
  }
}
