// fe-filepush/src/melt/vos.sdk.js
/**
 *  @file: vos.sdk.js is a <js> file
 *  @date: File created 2017-09-29 17:11
 *  @author:  shangwenhe
 */

'use strict';
import OSS from './vendorsdk/aliyun/aliyun.entry';
import KODO from './vendorsdk/qiniu/qiniu.entry';


/**
 * @class vos
 *    @params info
 * @function put
 *    @params files
 */

class vos {
  constructor(info) {
    switch (info.cloud) {
      case 'oss':
        let token = JSON.parse(info.token);
        // openclud opcld 开放云
        this.opcld = new OSS({
          region: info.region,
          //云账号AccessKey有所有API访问权限，建议遵循阿里云安全最佳实践，创建并使用STS方式来进行API访问
          accessKeyId: token.accessKeyId,
          accessKeySecret: token.accessKeySecret,
          stsToken: token.securityToken,
          endpoint: info.endpoint,
          bucket: info.bucket
        });
        break;
      case 'kodo':
        // 七牛云
        this.opcld = new KODO({
          bucket: info.bucket,
          // key: info.key,
          token: info.token
        });
        break;
      default:
        throw new Error('无法识别第三方云');
    }
  }

  /**
   * @desc 上传文件
   * @params files 所要上传的文件
   */
  put(files, callback, process) {
    this.opcld.put(files, callback, process);
  }

  /**
   * @desc 上传完成后通知服务端分发文件
   */
	notifier(info, callback) {
		callback(info);
  }
}

export default vos;
